orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,550 @@
+                                      concat_output_args, &attributes, kOnnxDomain);
+    ORT_ENFORCE(graph.SetOpSchemaFromRegistryForNode(concat_node), "Failed to concat shape for " + concat_node.Name());
+    concat_node.SetExecutionProviderType(node.GetExecutionProviderType());
+    unflattened_shape_arg = concat_output_args[0];
+  }
+
+  InlinedVector<NodeArg*> gathergrad_input_args;
+  gathergrad_input_args.reserve(3);
+  gathergrad_input_args.push_back(new_shape_arg);
+  gathergrad_input_args.push_back(gathergrad_index_arg);
+  gathergrad_input_args.push_back(node.MutableInputDefs()[in_index]);
+
+  InlinedVector<NodeArg*> gathergrad_output_args;
+  gathergrad_output_args.push_back(
+      &graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("padding_recover_result"),
+                                nullptr));
+
+  Node* new_gathergrad_node = onnxruntime::optimizer::compute_optimizer::InsertIntermediateNodeOnDestInput(
+      graph, node,
+      in_index,
+      2,
+      0,
+      graph.GenerateNodeName("PaddingRecover"),
+      "GatherGrad",
+      "GatherGrad node to recover invalid tokens.",
+      gathergrad_input_args,
+      gathergrad_output_args,
+      {},
+      kMSDomain,
+      logger);
+
+  new_gathergrad_node->SetExecutionProviderType(node.GetExecutionProviderType());
+  auto gathergrad_out_arg = new_gathergrad_node->MutableOutputDefs()[0];
+
+  InlinedVector<NodeArg*> reshape_input_args;
+  reshape_input_args.push_back(gathergrad_out_arg);
+  reshape_input_args.push_back(unflattened_shape_arg);
+  InlinedVector<NodeArg*> reshape_output_args{&graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("reshape_result"),
+                                                                        nullptr)};
+  Node* new_reshape_node = onnxruntime::optimizer::compute_optimizer::InsertIntermediateNodeOnDestInput(
+      graph, node,
+      in_index,
+      0,
+      0,
+      graph.GenerateNodeName("RecoverShape"),
+      "Reshape",
+      "Reshape node to recover invalid tokens.",
+      reshape_input_args,
+      reshape_output_args,
+      {},
+      kOnnxDomain,
+      logger);
+  new_reshape_node->SetExecutionProviderType(node.GetExecutionProviderType());
+  return new_reshape_node->MutableOutputDefs()[0];
+}
+
+Status PaddingElimination::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  LOG_DEBUG_INFO(logger, "Enter PaddingElimination");
+
+  if (sparse_embedding_input_names_.size() == 0) {
+    LOG_DEBUG_INFO(logger, "Exit PaddingElimination, no sparse embedding input names.");
+    return Status::OK();
+  }
+
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  Node* embedding_node = nullptr;
+  NodeArg* input_ids_arg = nullptr;
+  // Make sure each egde in candidate_edges has two consecutive dims to be flattened.
+  // Now only support the case that the first two dims are consecutive.
```

Comment:
a comment including an example graph (before and after) might be helpful to get the idea quickly.